### DISCUSSION TITLE
[Proposal][Generic] Inference item type in collection

### DISCUSSION THREAD

**Discussion by MickaelThumerel** [dotnet/roslyn#76516] [Ideas]
**Proposition**: Allow collection item type decomposition during inference

In C# to improve readability generic value are often automatically detected by the system "Inference of type"

Example:

A Method:
```csharp
public  Task Inject<TType>(TType object);
```

Usage:
```csharp
string test = "42";

Inject(test); // auto-inference TType to be a string
Inject(42); // auto-inference TType to be an int
```

**Issues**:

In the following example you find a 2 generics parameters one in the collection type the other represent the item in this collection.

```csharp
public  Task<TItemType> FirstInjectCollection<TCollectionType, TItemType>(TCollectionType object)
            where TCollectionType : IReadOnlyCollection<TItemType>;
```

The problem is that if type's interference doesn't solved all the parameter then you need to write manually all of them.

Resulting :
```csharp
List<string> test;
var firstItem = FirstInjectCollection<List<string>, string>(test)
```

It's heavy to ready.
A solution exist, using kind of placeholder parameter to ensure the resolution:

```csharp
public  Task<TItemType> FirstInjectCollection<TCollectionType, TItemType>(TCollectionType object, TItemType _)
            where TCollectionType : IReadOnlyCollection<TItemType>;
```

Resulting of a usage :

```csharp
List<string> test;
var firstItem = FirstInjectCollection(test, (string?)null)
```

Could introduce miss comprehension and possible bug to have a null value inject this way.

**Proposition**:

Enhance inference to be able to decompose **SIMPLE** identified .net type like, IEnumerable, IReadOnlyCollection, IReadOnlyDictionary, Span ...

Resulting:

```csharp
public  Task<TItemType> FirstInjectCollection<TCollectionType, TItemType>(TCollectionType object)
            where TCollectionType : IReadOnlyCollection<TItemType>;
            
Usage:

List<string> test;
var firstItem = FirstInjectCollection(test)
```

**Comment by huoyaoyuan** [comment]
Proposal for language goes to https://github.com/dotnet/csharplang/discussions . This repo is for compiler implementation.

There are already proposals to improve type inference like https://github.com/dotnet/csharplang/discussions/7467 .
